packages/aws-cdk/lib/api/util/checks.ts

```diff
@@ -0,0 +1,88 @@
+import { debug } from '../../logging';
+import { ISDK } from '../aws-auth/sdk';
+
+export async function determineAllowCrossAccountAssetPublishing(sdk: ISDK, customStackName?: string): Promise<boolean> {
+  const stackName = customStackName || 'CDKToolkit';
+  if (!(await _stagingBucketInToolkitStack(sdk, stackName))) {
+    // indicates an intentional cross account setup
+    return true;
+  }
+
+  if (await _bootstrapVersion(sdk, stackName) >= 21) {
+    // bootstrap stack version 21 contains a fix that will prevent cross
+    // account publishing on the IAM level
+    // https://github.com/aws/aws-cdk/pull/30823
+    return true;
+  }
+
+  // other scenarios are highly irregular and potentially dangerous so we prevent it by
+  // instructing cdk-assets to detect foriegn bucket ownership and reject.
+  return false;
+}
+
+async function _stagingBucketInToolkitStack(sdk: ISDK, stackName?: string): Promise<boolean> {
+  const toolkitStackName = stackName || 'CDKToolkit';
+  try {
+    // Create the SDKs for this enviornment, and use it to get the stack
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+    const response = await cfn.describeStacks({ StackName: toolkitStackName }).promise();
+
+    // Check if the toolkit stack exists
+    if (!response.Stacks || response.Stacks.length === 0) {
+      throw new Error(`Toolkit stack ${toolkitStackName} not found`);
```

Comment:
Throwing errors feels a bit dangerous here. Are these going to be swallowed?

Reply:
They didn't always get swallowed but I agree they should so I'm gonna swallow them now